model/labels/regexp.go

```diff
@@ -91,60 +90,91 @@ func findSetMatches(re *syntax.Regexp, base string) []string {
 		// In some case like negation [^0-9] a lot of possibilities exists and that
 		// can create thousands of possible matches at which points we're better off using regexp.
 		if totalSet > maxSetMatches {
-			return nil
+			return nil, false
 		}
 		for i := 0; i+1 < len(re.Rune); i = i + 2 {
 			lo, hi := re.Rune[i], re.Rune[i+1]
 			for c := lo; c <= hi; c++ {
 				matches = append(matches, base+string(c))
 			}
-
 		}
-		return matches
+		return matches, isCaseSensitive(re)
 	default:
-		return nil
+		return nil, false
 	}
-	return nil
+	return nil, false
 }
 
-func findSetMatchesFromConcat(re *syntax.Regexp, base string) []string {
+func findSetMatchesFromConcat(re *syntax.Regexp, base string) (matches []string, caseSensitive bool) {
 	if len(re.Sub) == 0 {
-		return nil
+		return nil, false
 	}
 	clearCapture(re.Sub...)
-	matches := []string{base}
+
+	matches = []string{base}
+	var matchesCaseSensitive *bool
 
 	for i := 0; i < len(re.Sub); i++ {
 		var newMatches []string
 		for _, b := range matches {
-			m := findSetMatches(re.Sub[i], b)
+			m, caseSensitive := findSetMatches(re.Sub[i], b)
 			if m == nil {
-				return nil
+				return nil, false
 			}
 			if tooManyMatches(newMatches, m...) {
-				return nil
+				return nil, false
+			}
+
+			// All matches must have the same case sensitivity. If it's the first set of matches
+			// returned, we store its sensitivity as the expected case, and then we'll check all
+			// other ones.
+			if matchesCaseSensitive == nil {
+				matchesCaseSensitive = &caseSensitive
 			}
+			if *matchesCaseSensitive != caseSensitive {
+				return nil, false
+			}
```

Comment:
Very opinonated, but IMO just checking indexes `j > 0` (with j being the index of `b`) reads easier than pointer setting:

```suggestion
			if j > 0 && matchesCaseSensitive != caseSensitive {
				return nil, false
			}
			matchesCaseSensitive = caseSensitive
```

Feel free to ignore this.
_(Same applied to the `findSetMatchesFromAlternate` below)_

Reply:
We can't just check `j` but we also have to check `i`. Anyway, I committed a slightly modified version of your idea:
https://github.com/grafana/mimir-prometheus/pull/430/commits/8008a119fbb13edda7074ddf0aecf1f62d050491

WDYT?